src/nlp_expr.jl

```diff
@@ -0,0 +1,330 @@
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+struct NonlinearExpr <: AbstractJuMPScalar
+    head::Symbol
+    args::Vector{Any}
+end
+
+# We include this method so that we can refactor the internal representation of
+# NonlinearExpr without having to rewrite the method overloads.
+function NonlinearExpr(head::Symbol, args...)
+    return NonlinearExpr(head, Any[args...])
+end
+
+Base.length(x::NonlinearExpr) = length(x.args)
+Base.getindex(x::NonlinearExpr, i::Int) = x.args[i]
+
+function function_string(::MIME"text/plain", x::NonlinearExpr)
+    io, stack, is_open = IOBuffer(), Any[x], true
+    while !isempty(stack)
+        arg = pop!(stack)
+        if !is_open && arg != ')'
+            print(io, ", ")
+        end
+        if arg isa NonlinearExpr
+            print(io, arg.head, "(")
+            push!(stack, ')')
+            for i in length(arg):-1:1
+                push!(stack, arg[i])
+            end
+        else
+            print(io, arg)
+        end
+        is_open = arg isa NonlinearExpr
+    end
+    seekstart(io)
+    return read(io, String)
+end
+
+function function_string(::MIME"text/latex", expr::NonlinearExpr)
+    return "\\textsf{$(function_string(MIME("text/plain"), expr))}"
+end
+
+_isequal(x, y) = x == y
+_isequal(x::T, y::T) where {T<:AbstractJuMPScalar} = isequal_canonical(x, y)
+
+function isequal_canonical(x::NonlinearExpr, y::NonlinearExpr)
+    return x.head == y.head &&
+           length(x) == length(y) &&
+           all(i -> _isequal(x[i], y[i]), 1:length(x))
+end
+
+function MOI.Nonlinear.parse_expression(
+    data::MOI.Nonlinear.Model,
+    expr::MOI.Nonlinear.Expression,
+    x::NonlinearExpr,
+    parent::Int,
+)
+    stack = Tuple{Int,Any}[(parent, x)]
+    while !isempty(stack)
+        parent_node, arg = pop!(stack)
+        if arg isa NonlinearExpr
+            _parse_without_recursion_inner(stack, data, expr, arg, parent_node)
+        else
+            # We can use recursion here, because NonlinearExpr only occur in
+            # other NonlinearExpr.
+            MOI.Nonlinear.parse_expression(data, expr, arg, parent_node)
+        end
+    end
+    return
+end
+
+function _parse_without_recursion_inner(stack, data, expr, x, parent)
+    id = get(data.operators.univariate_operator_to_id, x.head, nothing)
+    node_type = if length(x) == 1 && id !== nothing
+        MOI.Nonlinear.NODE_CALL_UNIVARIATE
+    else
+        id = get(data.operators.multivariate_operator_to_id, x.head, nothing)
+        @assert id !== nothing
+        MOI.Nonlinear.NODE_CALL_MULTIVARIATE
+    end
+    push!(expr.nodes, MOI.Nonlinear.Node(node_type, id, parent))
+    parent = length(expr.nodes)
+    for i in length(x):-1:1  # Args need to be pushed onto the stack in reverse
+        push!(stack, (parent, x[i]))
+    end
+    return
+end
+
+# Method definitions
+
+Base.zero(::Type{NonlinearExpr}) = NonlinearExpr(:+, 0.0)
```

Comment:
just wondering - how often are these zero/one methods likely to be called? since this representation seems a bit inefficient (but keeps things simple)